File: Python/Problem 3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen: Dict[str, int] = {}
        max_length = 0
        start = 0

        for end, char in enumerate(s):
            if char in seen and seen[char] >= start:
                start = seen[char] + 1
            seen[char] = end
            max_length = max(max_length, end - start + 1)

        return max_length

File: Java/Problem 3.java
import java.util.HashMap;

public class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> seen = new HashMap<>();
        int maxLength = 0;
        int start = 0;

        for (int end = 0; end < s.length(); end++) {
            char currentChar = s.charAt(end);
            if (seen.containsKey(currentChar) && seen.get(currentChar) >= start) {
                start = seen.get(currentChar) + 1;
            }
            seen.put(currentChar, end);
            maxLength = Math.max(maxLength, end - start + 1);
        }

        return maxLength;
    }
}
